store/src/java/com/zimbra/cs/service/mail/SendShareNotification.java

```diff
@@ -745,5 +754,31 @@ public void run() {
         senderThread.start();
     }
 
+    private void processDBFilesSharedWithMe(MailItem reqItem, Action action, OperationContext octxt, String granteeId,
+            String fileName, String ownerAccountId, int ownerFileId, String fileUUID, String fileOwnerName,
+            String rights) throws ServiceException {
+        Mailbox mbox = MailboxManager.getInstance().getMailboxByAccountId(granteeId);
+        // if file is shared, this file should be available at receivers end
+        if (reqItem.getType().equals(Type.DOCUMENT) && action == null) {
+            Document docreqItem = (Document) reqItem;
+            mbox.createFileSharedWithMe(mbox, octxt, FolderConstants.ID_FOLDER_FILE_SHARED_WITH_ME, fileName,
+                    ownerAccountId, ownerFileId, fileUUID, fileOwnerName, docreqItem.getContentType(),
+                    docreqItem.getSize(), rights);
+        } // if file access is revoked at source delete it from receiver end as well
+        else if (reqItem.getType().equals(Type.DOCUMENT) && action != null && action.name().equals(REVOKE)) {
+            String fileUUIDToDelete = fileUUID + ":" + ownerFileId;
+            mbox.deleteFileSharedWithMe(mbox, octxt, fileUUIDToDelete);
+        } // if file information is updated, then update at receivers end as well
+          // the reason for delete and create is, upon executing update query the latest
+          // information wasn't returned
+          // on Search response as items were returned from cached data
```

Comment:
cache should be removed rather than recreating an item if possible. DB write process is higher cost on performance.